haystack/components/converters/openapi_functions.py

```diff
@@ -0,0 +1,177 @@
+        if not spec_version:
+            raise ValueError(f"Invalid OpenAPI spec provided. Could not extract version from {service_openapi_spec}")
+        service_openapi_spec_version = int(spec_version.split(".")[0])
+        min_required_version = 3
+
+        # Compare the versions
+        if service_openapi_spec_version < min_required_version:
+            raise ValueError(f"Invalid OpenAPI spec version {service_openapi_spec_version}. Must be at least 3.0.0")
+
+        functions: List[Dict[str, Any]] = []
+        for path_methods in service_openapi_spec["paths"].values():
+            for method_specification in path_methods.values():
+                resolved_spec = jsonref.replace_refs(method_specification)
+                if isinstance(resolved_spec, dict):
+                    function_name = resolved_spec.get("operationId")
+                    desc = resolved_spec.get("description") or resolved_spec.get("summary", "")
+
+                    schema = {"type": "object", "properties": {}}
+
+                    req_body = (
+                        resolved_spec.get("requestBody", {})
+                        .get("content", {})
+                        .get("application/json", {})
+                        .get("schema")
+                    )
+                    if req_body:
+                        schema["properties"]["requestBody"] = req_body  # type: ignore
+
+                    params = resolved_spec.get("parameters", [])
+                    if params:
+                        param_properties = {param["name"]: param["schema"] for param in params if "schema" in param}
+                        schema["properties"]["parameters"] = {"type": "object", "properties": param_properties}  # type: ignore
+
+                    # these three fields are minimal requirement for OpenAI function calling
+                    if function_name and desc and schema:
+                        functions.append({"name": function_name, "description": desc, "parameters": schema})
+                    else:
+                        logger.warning(
+                            "Invalid OpenAPI spec format provided. Could not extract function from %s", resolved_spec
+                        )
+
+                else:
+                    logger.warning(
+                        "Invalid OpenAPI spec format provided. Could not extract function from %s", resolved_spec
+                    )
+
+        return functions
+
+    def _parse_openapi_spec(self, content: str) -> Dict[str, Any]:
+        """
+        Parses OpenAPI specification content, supporting both JSON and YAML formats.
+
+        :param content: The content of the OpenAPI specification.
+        :type content: str
+        :return: The parsed OpenAPI specification.
+        :rtype: Dict[str, Any]
+        """
+        try:
+            return json.loads(content)
+        except json.JSONDecodeError:
+            return yaml.safe_load(content)
```

Comment:
I would change this function. 
If there is actually something wrong with the JSON string the user is going to get an error about YAML loading error. That would be extremely confusing.

Reply:
Roght, good point!